README.md
# Lockit delete account

[![Build Status](https://travis-ci.org/zemirco/lockit-delete-account.svg?branch=master)](https://travis-ci.org/zemirco/lockit-delete-account)
[![NPM version](https://badge.fury.io/js/lockit-delete-account.svg)](http://badge.fury.io/js/lockit-delete-account)
[![Dependency Status](https://david-dm.org/zemirco/lockit-delete-account.svg)](https://david-dm.org/zemirco/lockit-delete-account)

Delete user accounts in your Express app. The module is part of [Lockit](https://github.com/zemirco/lockit).

## Installation

`npm install lockit-delete-account`

```js
var DeleteAccount = require('lockit-delete-account');
var utils = require('lockit-utils');
var config = require('./config.js');

var db = utils.getDatabase(config);
var adapter = require(db.adapter)(config);

var app = express();

// express settings
// ...
// sessions are required - either cookie or some sort of db
app.use(cookieParser());
app.use(cookieSession({
  secret: 'this is my super secret string'
}));

// create new DeleteAccount instance
var deleteAccount = new DeleteAccount(config, adapter);

// use deleteAccount.router with your app
app.use(deleteAccount.router);
```

## Configuration

More about configuration at [Lockit](https://github.com/zemirco/lockit).

## Features

 - input validation
 - match public phrase
 - session verification
 - kill current session
 - remove user from db

## Routes included

 - GET /delete-account
 - POST /delete-account

## REST API

If you've set `exports.rest` in your `config.js` the module behaves as follows.

 - all routes have `/rest` prepended
 - `GET /rest/delete-account` is `next()`ed and you can catch `/delete-account` on the client
 - `POST /rest/delete-account` stays the same but sends JSON

## Test

`grunt`

## License

MIT


History.md

##### 1.0.5 / 2015-07-06

- update dependencies

##### 1.0.4 / 2015-07-03

- use eslint
- update dependencies
- use make instead grunt

##### 1.0.3 / 2014-05-27

- set `autocomplete="off"`
- use Bootstrap conditional classes
- use `utils.destroy` to kill session
- update dependencies

##### 1.0.2 / 2014-05-23

- update lockit-utils

##### 1.0.1 / 2014-05-23

- refactor code
- update dependencies

##### 1.0.0 / 2014-04-19

- requires Express 4.x
- makes use of `express.Router()`. No need to pass `app` around as argument.

  **old**

  ```js
  var DeleteAccount = require('lockit-delete-account');

  var deleteAccount = new DeleteAccount(app, config, adapter);
  ```

  **new**

  ```js
  var DeleteAccount = require('lockit-delete-account');

  var deleteAccount = new DeleteAccount(config, adapter);
  app.use(deleteAccount.router);
  ```

- proper Error handling. All Errors are piped to next middleware.

  **old**

  ```js
  if (err) console.log(err);
  ```

  **new**

  ```js
  if (err) return next(err);
  ```

  Make sure you have some sort of error handling middleware at the end of your
  routes (is included by default in Express 4.x apps if you use the `express-generator`).

##### 0.6.0 / 2014-04-11

- `username` becomes `name`


test/app.js

/**
 * Module dependencies.
 */

var express = require('express');
var routes = require('./routes');
var user = require('./routes/user');
var http = require('http');
var path = require('path');

// require delete account middleware
var deleteAccount = require('../index.js');

function start(config) {

  config = config || require('./config.js');
  
  var app = express();

// set basedir so views can properly extend layout.jade
  app.locals.basedir = __dirname + '/views';

// all environments
  app.set('port', process.env.PORT || config.port || 3000);
  app.set('views', __dirname + '/views');
  app.set('view engine', 'jade');
  // make JSON output simpler for testing
  app.set('json spaces', 0);
  app.use(express.favicon());
  app.use(express.bodyParser());
  app.use(express.methodOverride());
  app.use(express.cookieParser('your secret here'));
  app.use(express.cookieSession());

// set a dummy session for testing purpose
  app.use(function(req, res, next) {
    req.session.username = 'john';
    req.session.email = 'john@email.com';
    next();
  });

// use delete account middleware with testing options
  deleteAccount(app, config);

  app.use(app.router);
  app.use(express.static(path.join(__dirname, 'public')));

// development only
  if ('development' == app.get('env')) {
    app.use(express.errorHandler());
  }

  app.get('/', routes.index);
  app.get('/users', user.list);

  http.createServer(app).listen(app.get('port'), function(){
    console.log('Express server listening on port ' + app.get('port'));
  });
  
  return app;
  
}

// export app for testing
if(require.main === module){
  // called directly
  start();
} else {
  // required as a module -> from test file
  module.exports = function(config) {
    return start(config);
  };
}

test/csrf.js

var request = require('supertest');
var should = require('should');
var cookie = require('cookie');
var utls = require('lockit-utils');

var config = require('./app/config.js');
var app = require('./app/app.js');

var db = utls.getDatabase(config);
var adapter = require(db.adapter)(config);

var _config = JSON.parse(JSON.stringify(config));
_config.port = 7000;
_config.csrf = true;
var _app = app(_config);

describe('# csrf', function() {

  describe('GET /delete-account', function() {

    it('should include the token in the view', function(done) {
      request(_app)
        .get('/delete-account')
        .end(function(err, res) {
          var cookies = cookie.parse(res.headers['set-cookie'][0]);
          var token = cookies.csrf;
          res.text.should.containEql('name="_csrf" value="' + token + '"');
          res.statusCode.should.equal(200);
          done();
        });
    });

  });

});


test/rest.js
var request = require('supertest');
var should = require('should');
var utls = require('lockit-utils');

var config = require('./app/config.js');
var app = require('./app/app.js');

var db = utls.getDatabase(config);
var adapter = require(db.adapter)(config);

var _config = JSON.parse(JSON.stringify(config));
_config.port = 5000;
_config.rest = true;
var _app = app(_config);

describe('# rest enabled', function() {

  // add a dummy user to db
  before(function(done) {
    adapter.save('john', 'john@email.com', 'password', done);
  });

  describe('GET /delete-account', function() {

    it('should render the default route (REST)', function(done) {
      request(_app)
        .get('/rest/delete-account')
        .end(function(err, res) {
          res.statusCode.should.equal(404);
          done();
        });
    });

  });

  describe('POST /delete-account', function() {

    it('should show an error message when an input field is empty (REST)', function(done) {
      request(_app)
        .post('/rest/delete-account')
        .send({name: '', phrase: 'lorem', password: 'secret'})
        .end(function(error, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"All fields are required"}');
          done();
        });
    });

    it('should show an error message when phrase is incorrect', function(done) {
      request(_app)
        .post('/rest/delete-account')
        .send({name: 'john', phrase: 'please do not delete my account forever', password: 'secret'})
        .end(function(error, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"Phrase doesn\'t match"}');
          done();
        });
    });

    it('should show an error message when session doesn\'t match username', function(done) {
      request(_app)
        .post('/rest/delete-account')
        .send({name: 'jack', phrase: 'please delete my account forever', password: 'secret'})
        .end(function(error, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"Please enter your username"}');
          done();
        });
    });

    it('should show an error message when password is incorrect', function(done) {
      request(_app)
        .post('/rest/delete-account')
        .send({name: 'john', phrase: 'please delete my account forever', password: 'secret'})
        .end(function(error, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"Password is wrong"}');
          done();
        });
    });

    it('should delete a user from db when everything is fine', function(done) {
      request(_app)
        .post('/rest/delete-account')
        .send({name: 'john', phrase: 'please delete my account forever', password: 'password'})
        .end(function(error, res) {
          res.statusCode.should.equal(204);
          done();
        });
    });

  });

});


index.js
'use strict';

var path = require('path');
var events = require('events');
var util = require('util');
var express = require('express');
var pwd = require('couch-pwd');
var utils = require('lockit-utils');



/**
 * Internal helper functions
 */
function join(view) {
  return path.join(__dirname, 'views', view);
}



/**
 * DeleteAccount constructor function.
 *
 * @constructor
 * @param {Object} config
 * @param {Object} adapter
 */
var DeleteAccount = module.exports = function(config, adapter) {

  if (!(this instanceof DeleteAccount)) {return new DeleteAccount(config, adapter); }

  this.config = config;
  this.adapter = adapter;

  // call super constructor function
  events.EventEmitter.call(this);

  // set default route
  var route = config.deleteAccount.route || '/delete-account';

  // add prefix when rest is active
  if (config.rest) {route = '/rest' + route; }

  /**
   * Routes
   */

  var router = new express.Router();
  router.get(route, utils.restrict(config), this.getDelete.bind(this));
  router.post(route, utils.restrict(config), this.postDelete.bind(this));
  this.router = router;

};

util.inherits(DeleteAccount, events.EventEmitter);



/**
 * GET /delete-account.
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
DeleteAccount.prototype.getDelete = function(req, res, next) {
  var config = this.config;

  // do not handle the route when REST is active
  if (config.rest) {return next(); }

  // custom or built-in view
  var view = config.deleteAccount.views.remove || join('get-delete-account');

  res.render(view, {
    title: 'Delete account',
    basedir: req.app.get('views')
  });
};



/**
 * POST /delete-account.
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
DeleteAccount.prototype.postDelete = function(req, res, next) {
  var config = this.config;
  var adapter = this.adapter;
  var that = this;

  // verify input fields
  var name = req.body.name;
  var phrase = req.body.phrase;
  var password = req.body.password;

  var error = null;

  // check for valid inputs and valid session
  if (!name || !phrase || !password) {
    error = 'All fields are required';
  } else if (phrase !== 'please delete my account forever') {
    error = 'Phrase doesn\'t match';
  } else if (req.session.name !== name) {
    error = 'Please enter your username';
  }

  // custom or built-in view
  var view = config.deleteAccount.views.remove || join('get-delete-account');

  if (error) {
    // do not handle the route when REST is active
    if (config.rest) {return res.json(403, {error: error}); }

    res.status(403);
    return res.render(view, {
      title: 'Delete account',
      error: error,
      basedir: req.app.get('views')
    });
  }

  // get user from db
  adapter.find('name', name, function(err, user) {
    if (err) {return next(err); }

    // no need to check if user exists in db since we are already checking against current session

    // if user comes from couchdb it has an 'iterations' key
    if (user.iterations) {pwd.iterations(user.iterations); }

    // verify user password
    pwd.hash(password, user.salt, function(hashErr, hash) {
      if (hashErr) {return next(hashErr); }

      // compare hash with hash from db
      if (hash !== user.derived_key) {
        error = 'Password is wrong';

        // do not handle the route when REST is active
        if (config.rest) {return res.json(403, {error: error}); }

        res.status(403);
        return res.render(view, {
          title: 'Delete account',
          error: error,
          basedir: req.app.get('views')
        });

      }

      // delete user from db :(
      adapter.remove(name, function(removeErr) {
        if (removeErr) {return next(removeErr); }

        // kill session
        utils.destroy(req, function() {

          // emit 'delete' event
          that.emit('delete', user, res);

          if (config.deleteAccount.handleResponse) {

            // do not handle the route when REST is active
            if (config.rest) {return res.send(204); }

            view = config.deleteAccount.views.removed || join('post-delete-account');

            // render success message
            res.render(view, {
              title: 'Account deleted',
              basedir: req.app.get('views')
            });

          }

        });

      });

    });

  });
};
